xbmc/network/upnp/UPnPPlayer.cpp

```diff
@@ -525,11 +524,14 @@ void CUPnPPlayer::DoAudioWork()
                                                  static_cast<void*>(new CFileItem(*item)));
     }
 
-    NPT_CHECK_LABEL(m_delegate->m_transport->GetStateVariableValue("TransportState", data), failed);
-    if(data == "STOPPED")
     {
-      m_started = false;
-      m_callback.OnPlayBackEnded();
+      std::unique_lock<CCriticalSection> lock(m_delegate->m_section);
+      if (m_delegate->m_trainfo.cur_transport_state == "STOPPED")
+      {
+        m_logger->info("Transport state flagged as STOPPED. Triggering OnPlayBackEnded.");
+        m_started = false;
+        m_callback.OnPlayBackEnded();
```

Comment:
Is this a blocking callback?

Reply:
Yeah it is trapped by python and others.
I see your point, will move it out of the scoped lock 👍

Thanks

Reply:
Improved on the last revision. Runtime tested again.